tests/zfs-tests/cmd/zfs_diff-socket.c

```diff
@@ -37,6 +37,8 @@ main(int argc, char *argv[])
 	}
 	path = argv[1];
 	size =  sizeof (sock.sun_path);
+	/* Suppress deprecation warning */
+#undef strncpy
```

Comment:
Ouch. This is a bug in Clang.

```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#undef strncpy

int main()
{
        char buf[256];
        char hello[] = "hello world\n";

        strncpy(buf, hello, sizeof(buf));

        printf("%s", buf);

        return 0;
}
```

`gcc -D"strncpy(...)=__attribute__((deprecated(\"strncpy(3) is deprecated. Use strlcpy(3) instead\!\"))) strncpy(__VA_ARGS__)" hello.c` works fine. `clang -D"strncpy(...)=__attribute__((deprecated(\"strncpy(3) is deprecated. Use strlcpy(3) instead\!\"))) strncpy(__VA_ARGS__)" hello.c` breaks.

I am going to try working around this by disabling the deprecation warning when building with clang. I imagine that I will file a bug report with Clang later too.

Reply:
Autotools did not like my attempts to detect Clang. Rather than modify autotools to detect Clang, I converted the file to use `strlcpy()` and added this:

```
/* XXX: Workaround not linking to libspl */
#include "../../../lib/libspl/strlcpy.c"
```

That file uses an existing autotools check via  `#ifndef HAVE_STRLCPY` to determine whether to build `strlcpy()` and the result is passed to the build environment for that test. This allows us to use `strlcpy()` in the test on Linux without breaking the build on FreeBSD. :)

Reply:
I am not marking this conversation resolved mostly because I have not done anything with the Clang bug that I found yet.

Reply:
https://github.com/llvm/llvm-project/issues/57954

An upstream issue has been filed.